rust/kernel/of.rs

```diff
@@ -0,0 +1,67 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Devicetree and Open Firmware abstractions.
+//!
+//! C header: [`include/linux/of_*.h`](../../../../include/linux/of_*.h)
+
+use alloc::boxed::Box;
+
+use crate::{
+    bindings, c_types,
+    error::{Error, Result},
+    types::PointerWrapper,
+    CStr,
+};
+
+use core::mem::transmute;
+
+/// Wraps a kernel Open Firmware / devicetree match table.
+///
+/// Rust drivers may create this structure to match against devices
+/// described in the devicetree.
+///
+/// # Invariants
+///
+/// The last array element is always filled with zeros (the default).
+pub struct OfMatchTable(Box<[bindings::of_device_id; 2]>);
```

Comment:
I tried to support variable-size match tables by using const generics here, eg:
```rust
pub struct OfMatchTable<const N: usize>(Box<[bindings::of_device_id; N+1]>);
```
and everything appeared to work great, until building rustdoc. That failed with the internal compiler error described [here](https://github.com/rust-lang/rust/issues/77647).